Dafny 4: User guide updates (#3580)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

They are not repeated here to avoid replicating information that
easily becomes inconsistent and out of date.
The dafny tool can also be installed using `dotnet tool install --global dafny`
(presuming that `dotnet` is already installed on your system).

> Most users will find it most convenient to install the pre-built Dafny binaries available on the project release site or in dotnet.

Reviewer: ```suggestion
Most users will find it most convenient to install the pre-built Dafny binaries available on the project release site or using the `dotnet` CLI.
```

As is typical for Open Source projects, dafny can also be built directly from the source files maintained in the github project.

Current and past Dafny binary releases can be found at
[https://github.com/dafny-lang/dafny/releases](https://github.com/dafny-lang/dafny/releases) for each supported platform.
Each release is a .zip file with a name combining the release name and the

> platform. Current platforms are Windows 11, Ubuntu 20ff, and MacOS 10.14ff.

Reviewer: What do the "ff" suffixes mean?

Author: And following. But I revised it to be clearer. Besides ff has lots more meanings today. (friends forever?)

#### 13.5.1.4. `dafny translate <language>` {#sec-dafny-translate}

The `dafny translate` command translates Dafny source code to source code for another target programming language.

> The command always performs the actions of `dafny resolve` and, unless `--no-verify is an option`, does the actions of `dafny verify`.

Reviewer: ```suggestion
The command always performs the actions of `dafny resolve` and, unless the `--no-verify` option is specified, does the actions of `dafny verify`.
```

- 2 for parsing, type-checking or serious errors in running the auditor (e.g. failure to write a report or when report comparison fails)
- 0 for normal operation, including operation that identifies audit findings

It also takes the `--verbose` option, which then gives information about the files being formatted.

#### 13.5.1.9. `dafny format` {#sec-dafny-format}

> Dafny supports a formatter, which for now only changes the indentation lines in a Dafny file, so that it conforms

Reviewer: I prefer the original wording but it's not a blocker. Could we write "indentation of lines in..."?

Author: That is what I meant. Mikael's words say that every line is altered, which is not the case.

### 13.5.3. dafny exit codes {#sec-exit-codes}

> The basic dafny tool (resolve, verify, translate, build, run) terminates with these exit codes.

Reviewer: What does "basic" mean here?

Author: Revised

program during processing, including variations of the original source
code (which can be helpful for debugging).

* `--use-basename-for-filename` - when enabled, just the filename without the 
directory path is used in error messages; this make error message shorter and 
not tied to the local environment (which is a help in testing)

* `--output`, `-o` - location of output files [translate, build]

* `--show-snippets` - include with an error message some of the source code text
in the neighborhood of the error; the error location (file, line, column) is always given

* `--solver-log <file>` - [verification only] the file in which to place the SMT text sent to the solver

> * `--log-format <configuration` - [verification only] logs information about verification performance

Reviewer: ```suggestion
* `--log-format <configuration>` - [verification only] logs information about verification performance
```

Also, we should list the permitted `configuration`s.

> * `--log-format <configuration>` - [verification only] logs information about verification performance

Reviewer: Can we move the explanation of the various permitted configurations / file formats here (from the legacy options section)?

Author: I'll change this (as I did for other options) in a next PR.

disabled. Some of these options exist for backward compatibility with
older versions of Dafny.

* `--function-syntax` (value '3' or '4') - permits a choice of using the Dafny 3 syntax (`function` and `function method`)
or the Dafny 4 syntax (`ghost function` and `function`)

* `--quantifier-syntax` (value '3' or '4') - permits a choice between the Dafny 3 and Dafny 4 syntax for quantifiers

* `--unicode-char` - if false, the `char` type represents any UTF-16 code unit,
  that is, any 16-bit value, including surrogate code points and
  allows `\uXXXX` escapes in string and character literals.
  If true, `char` represnts any Unicode scalar value,
  that is, any Unicode code point excluding surrogates and
  allows `\U{X..X}` escapes in string and character literals. 
  The default is false for Dafny version 3 and true for version 4.

>   The legacy option was -unicodeChar:<n>` with values 0 and 1 for

Reviewer: ```suggestion
  The legacy option was `-unicodeChar:<n>` with values 0 and 1 for
```

  implicitly static and field declarations are not allowed at the
  module scope.

### 13.8.6. Controlling warnings {#sec-controlling-warnings}

These options control what warnings Dafny produces, and whether to treat
warnings as errors.

> * `--warn-as-errors (was `-warningsAsErrors`) - treat warnings as errors.

Reviewer: ```suggestion
* `--warn-as-errors` (was `-warningsAsErrors`) - treat warnings as errors.
```

> * `--warn-shadowing (was `-warnShadowing`) - emit a warning if the name 

Reviewer: ```suggestion
* `--warn-shadowing` (was `-warnShadowing`) - emit a warning if the name 
```

for verification. Options supported by Boogie on its own are also
supported by Dafny. Some of the Boogie options most relevant to Dafny
users include the following. We use the term "procedure" below to refer
to a Dafny function, lemma, method, or predicate, following Boogie
terminology.

* `--solver-path` - specifies a custom SMT solver to use

* `--solver-plugin` - specifies a plugin to use as the SMT solver, instead of an external pdafny translaterocess

* `--boogie-filter` - restricts the set of verification tasks (for debugging) 

* `--boogie` - arguments to send to boogie

Legacy options:

* `-proc:<name>` - verify only the procedure named `<name>`. The name
  can include `*` to indicate arbitrary sequences of characters.

* `-trace` - print extra information during verification, including
  timing, resource use, and outcome for each procedure incrementally, as
  verification finishes.

* `-randomSeed:<n>` - turn on randomization of the input that Boogie
  passes to the SMT solver and turn on randomization in the SMT solver
  itself.

  Certain Boogie inputs are unstable in the sense that changes to the
  input that preserve its meaning may cause the output to change. The

>   `-randomSeed`` option simulates meaning-preserving changes to the

Reviewer: ```suggestion
  `-randomSeed` option simulates meaning-preserving changes to the
```
